pkg/volume/azure_dd/azure_dd_block.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure_dd
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/util/mount"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/volumepathhandler"
+)
+
+var _ volume.VolumePlugin = &azureDataDiskPlugin{}
+var _ volume.PersistentVolumePlugin = &azureDataDiskPlugin{}
+var _ volume.BlockVolumePlugin = &azureDataDiskPlugin{}
+var _ volume.DeletableVolumePlugin = &azureDataDiskPlugin{}
+var _ volume.ProvisionableVolumePlugin = &azureDataDiskPlugin{}
+
+func (plugin *azureDataDiskPlugin) ConstructBlockVolumeSpec(podUID types.UID, volumeName, mapPath string) (*volume.Spec, error) {
+	pluginDir := plugin.host.GetVolumeDevicePluginDir(azureDataDiskPluginName)
+	blkutil := volumepathhandler.NewBlockVolumePathHandler()
+	globalMapPathUUID, err := blkutil.FindGlobalMapPathUUIDFromPod(pluginDir, mapPath, podUID)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(5).Infof("globalMapPathUUID: %s", globalMapPathUUID)
+
+	globalMapPath := filepath.Dir(globalMapPathUUID)
+	if len(globalMapPath) <= 1 {
+		return nil, fmt.Errorf("failed to get volume plugin information from globalMapPathUUID: %v", globalMapPathUUID)
+	}
+
+	return getVolumeSpecFromGlobalMapPath(globalMapPath)
+}
+
+func getVolumeSpecFromGlobalMapPath(globalMapPath string) (*volume.Spec, error) {
+	// Get volume spec information from globalMapPath
+	// globalMapPath example:
+	//   plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumeID}
+	//   plugins/kubernetes.io/azure-disk/volumeDevices/vol-XXXXXX
+	diskName := filepath.Base(globalMapPath)
+	if len(diskName) <= 1 {
+		return nil, fmt.Errorf("failed to get diskName from global path=%s", globalMapPath)
+	}
+	glog.V(5).Infof("got diskName(%s) from globalMapPath: %s", globalMapPath, diskName)
+	block := v1.PersistentVolumeBlock
+	pv := &v1.PersistentVolume{
+		Spec: v1.PersistentVolumeSpec{
```

Comment:
I found some plugins setting volumeName as Name in ObjectMeta on ConstructBlockVolumeSpec() call.
# local, iscsi, and fc set it, but aws_ebs, rbd, and gce don't. 
Should it be set here?

Reply:
No, here it's inside `ConstructBlockVolumeSpec` func:
```
	// ConstructBlockVolumeSpec constructs a volume spec based on the given
	// pod name, volume name and a pod device map path.
	// The spec may have incomplete information due to limited information
	// from input. This function is used by volume manager to reconstruct
	// volume spec by reading the volume directories from disk.
```
it does not have suffient `volumeName` value, e.g. following path only has disk name other than volumeName, while for local, iscsi, and fc, `ConstructBlockVolumeSpec` already provides `volumeName` parameter in the func call.
```
/var/lib/kubelet/plugins/kubernetes.io/azure-disk/volumeDevices/block-azuredisk-test
```

Reply:
Sorry for making you confuse.
I meant to say that we will be able to set volumeName to Name in metav1.ObjectMeta in getVolumeSpecFromGlobalMapPath(),
if we pass volumeName to getVolumeSpecFromGlobalMapPath() from ConstructBlockVolumeSpec().
(It seems that ConstructBlockVolumeSpec() allows the resulting spec to be incomplete, however, it would be better to try to make more complete.)

I would like to hear opinions from authors of other plugins whether above makes sense,
for the same logic can also be applied to RBD and GCE to make the resulting spec similar to that for local.

Local: @dhirajh
RBD: @sbezverk
GCE: @screeley44

Reply:
@mkimuram thanks, I have pushed a new commit to address this: https://github.com/kubernetes/kubernetes/pull/63841/commits/3683f3f1ecc26357a4b2cf42f330433fa0cb2190, PTAL

Reply:
LGTM.